components/camel-univocity-parsers/src/main/java/org/apache/camel/dataformat/univocity/Marshaller.java

```diff
@@ -86,7 +88,9 @@ private void writeRow(Exchange exchange, Object row, W writer) throws NoTypeConv
         Map<?, ?> map = convertToMandatoryType(exchange, Map.class, row);
         if (adaptHeaders) {
             for (Object key : map.keySet()) {
-                headers.add(convertToMandatoryType(exchange, String.class, key));

```

Comment:
And here I would create another List with the modified headers. Something like:

    List<String> headersToUse = new ArrayList(this.headers);
    headersToUse.add(convertToMandatoryType(exchange, String.class, key));

and use `headersToUse` after that instead of `this.headers`